Add proposal-dynamic-import docs (#2740)
babel/website: docs/plugin-proposal-dynamic-import.md

---
id: babel-plugin-proposal-dynamic-import
title: "@babel/plugin-proposal-dynamic-import"
sidebar_label: dynamic-import
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> > **NOTE**: This plugin is included in `@babel/preset-env`, in [ES2020](https://github.com/tc39/proposals/blob/master/finished-proposals.md).

Reviewer: NIT: This can be admonition?

Author: It could be. Here I align the styles to other plugins: https://github.com/babel/website/blob/ea7ed1ee28638f4b32164cdebedde02f05b0bd69/docs/plugin-proposal-private-methods.md?plain=1#L7

Transforms `import()` expressions to non-ESM module formats.

## When (not) to use this plugin

> If you are using Babel in production, you should normally not use this plugin. Instead, you should use
> a bundler, such as Webpack, Rollup or Parcel to handle `import()` expressions. However, there are still some valid use cases:

Reviewer 2: ```suggestion
If you are using a bundler, such as Webpack, Rollup or Parcel, you should **not** use this plugin and let your bundler handle `import()` expressions.

You should use this plugin if:
```

all the listed use cases can be in production, and many people use bundlers in development :stuck_out_tongue:

- You are authoring a Node.js library in ESM but want to distribute it in CommonJS(CJS): Install this plugin and [`@babel/plugin-transform-modules-commonjs`](./plugin-transform-modules-commonjs.md)
- You use [RequireJS](https://requirejs.org) to load modules in the browser: Install this plugin and [`@babel/plugin-transform-modules-amd`](./plugin-transform-modules-amd.md)

> - You use [SystemJS](https://github.com/systemjs/systemjs) to load modules in the browser: Install this plugin and [`@babel/plugin-transform-modules-systemjs`](./plugin-transform-modules-systemjs.md)

Author: Although currently `@babel/plugin-transform-modules-systemjs` can transform `import()` without this plugin, we will require it in Babel 8. So here we suggest using the plugin with the systemjs transform.

## Example
```js title="input.js"
import("jquery").then($ => {});
```

will be transformed to

<Tabs>
  <TabItem value="commonjs" label="CommonJS" default>

  ```js title="output.js"
  Promise.resolve()
    .then(() => _interopRequireWildcard(require("jquery")))
    .then(($) => {});
  ```
  </TabItem>
  <TabItem value="amd" label="AMD">

  ```js title="output.js"
  define(["require"], function (_require) {
    new Promise((_resolve, _reject) =>
      _require(
        ["jquery"],
        (imported) => _resolve(_interopRequireWildcard(imported)),
        _reject
      )
    ).then(($) => {});
  });
  ```
  </TabItem>
  <TabItem value="systemjs" label="SystemJS">

  ```js title="output.js"
  System.register([], function (_export, _context) {
    "use strict";

    return {
      setters: [],
      execute: function () {
        _context.import("jquery").then(($) => {});
      }
    };
  });
  ```
  </TabItem>
</Tabs>

## Installation

```shell npm2yarn
npm install --save-dev @babel/plugin-proposal-dynamic-import
```

## Usage

### With a configuration file (Recommended)

```json title="babel.config.json"
{

>   "plugins": ["@babel/plugin-proposal-dynamic-import"]

Reviewer 2: Can we include also one of the modules plugins?
